### PR TITLE
[info] Add listitem.hasproperty, listitem.property(subtitlecount) and

### PR DESCRIPTION
**Pull request by enen92**
avoid setting the language property if no language could be detected

## Description
This is an alternative to https://github.com/xbmc/xbmc/pull/24867#issuecomment-2056390310:

- Provides `Listitem.HasProperty` infobool to check if a given property is set (thus a generic way to avoid false positives while looking into those cases where a string property is explicitly set to an empty string vs non-existent property)
- Add `Listitem.Property(SubtitleCount)` to have a way of getting the number of subtitles of a given item
- Avoid setting the language property for subtitles if the language could not be detected.

### CODE REVIEW

File: xbmc/video/VideoThumbLoader.cpp [hunk 1]
```diff
@@ -577,10 +577,16 @@ void CVideoThumbLoader::DetectAndAddMissingItemData(CFileItem &item)
     }
 
     // add subtitle language properties
-    for (int i = 1; i <= details.GetSubtitleStreamCount(); i++)
+    int subtitleCount = details.GetSubtitleStreamCount();
```
Comment (CrystalP) [inline xbmc/video/VideoThumbLoader.cpp:580]: const

File: xbmc/video/VideoThumbLoader.cpp [hunk 2]
```diff
@@ -577,10 +577,16 @@ void CVideoThumbLoader::DetectAndAddMissingItemData(CFileItem &item)
...
+    item.SetProperty("SubtitleCount", std::to_string(subtitleCount));
+    for (int i = 1; i <= subtitleCount; i++)
     {
       std::string index = std::to_string(i);
-      item.SetProperty("SubtitleLanguage." + index, details.GetSubtitleLanguage(i).c_str());
+      const std::string language = details.GetSubtitleLanguage(i);
+      if (!language.empty())
+      {
+        item.SetProperty("SubtitleLanguage." + index, language);
+      }
```
Comment (CrystalP) [inline xbmc/video/VideoThumbLoader.cpp:589]: index string not needed for blank language, but if kept should be const.
```suggestion
      const std::string language = details,GetSubtitleLanguage(i);
      if (!language.empty())
        item.SetProperty("SubtitleLanguage." + std::to_string(i), language);
```

**Comment by CrystalP** [comment]
For consistency, should audio language be treated the same?

**Comment by dcfou** [comment]
@enen92 , I have no time unfortunately to build and test it. However, this sounds good despite there is many changes here.
To my point of view, both ways are fine. Let's try this one ;)

**Comment by CrystalP** [comment]
Is there a reason for making the subs count a text property? rather than a listitem.subtitlecount of int type?
Seems to me that this way the skin has to convert back to number for comparisons.
And seems that so far listitem.property() is for "dynamic" attributes of variable number that cannot be defined as listitem.something

What about the audio language? Can be missing as well.

The most generic solution to distinguish missing from empty property could be a C++-like optional string type, in the form of a String.HasValue() function I guess, but no idea of the code involved.

File: xbmc/GUIInfoManager.cpp [hunk 3]
```diff
@@ -7174,6 +7189,7 @@ const infomap listitem_labels[]= {{ "thumb",            LISTITEM_THUMB },
                                   { "isvideoextra",     LISTITEM_ISVIDEOEXTRA },
                                   { "videoversionname", LISTITEM_VIDEOVERSION_NAME },
                                   { "hasvideoextras",   LISTITEM_HASVIDEOEXTRAS },
+                                  { "hasproperty",   LISTITEM_HASPROPERTY },
```
Comment (ksooo) [inline xbmc/GUIInfoManager.cpp:7192]: We should align the indentation with the other table entries.

File: xbmc/video/VideoThumbLoader.cpp [hunk 4]
```diff
@@ -577,10 +577,15 @@ void CVideoThumbLoader::DetectAndAddMissingItemData(CFileItem &item)
     }
 
     // add subtitle language properties
-    for (int i = 1; i <= details.GetSubtitleStreamCount(); i++)
+    const int subtitleCount = details.GetSubtitleStreamCount();
+    item.SetProperty("SubtitleCount", std::to_string(subtitleCount));
+    for (int i = 1; i <= subtitleCount; i++)
```
Comment (ksooo) [inline xbmc/video/VideoThumbLoader.cpp:582]: i++ -> ++i ?(coding guidelines)

**Comment by enen92** [comment]
> @CrystalP To my point of view, audio language can be tested differently : an audio stream may have no language defined and ListItem.AudioLanguage will be empty. But an audio has a codec and has channel(s). So, as I do, you can test audio channel (or audio codec) which should never be empty, then you have an audio : in your skin display first an unknown country flag and over it the right flag if audio language is not empty. This way, if it is empty, you will have the unknown country flag.
> 
> @enen92 ... as I write this remark, I just have an other idea : Subtitles have also a codec (pgs, txt -srt, ssa, ...), so why don't add this rather than the way we start ? This seems more consistent with what is done for video and audio streams. Codec empty = no subtitle, codec filled = subtitle => language ? country flag : unknown flag. We will need a ListItem.Property(SubtitleCodec.x)) too. What do you think ?

Hum I like the idea. The problem with the codec is that subtitles can be external and thus the codec (or the parser in this case) is only set when Kodi actually starts playing the video file and opens the subtitle file for reading. Doing this when filling the db for external subtitles is a no-go due to performance. However we can add properties to flag a given subtitle stream as external or muxed/internal. I added both properties in https://github.com/enen92/xbmc/commit/10aef21714a60a279e12438c5c3d6fcaf1b6ef1d

Would this work? Looks less hacky than what is being done on this PR.

**Comment by dcfou** [comment]
@enen92  This seems great :) This way, even if codec is not in db, we have the subtitle type.

**Pull request closed without merging**